api/client/batch_client_test.py

```diff
@@ -1,10 +1,184 @@
+        self.assertEqual(data_points[0][0]['value'], 40891)
+        self.assertEqual(data_points[0][0]['unit_id'], 14)
+        self.assertEqual(data_points[0][0]['reporting_date'], None)
+        self.assertEqual(data_points[0][1]['start_date'], '2018-01-01T00:00:00.000Z')
+        self.assertEqual(data_points[0][1]['end_date'], '2018-12-31T00:00:00.000Z')
+        self.assertEqual(data_points[0][1]['value'], 56789)
+        self.assertEqual(data_points[0][1]['unit_id'], 14)
+        self.assertEqual(data_points[0][1]['reporting_date'], '2019-03-14T00:00:00.000Z')
+
+    def test_batch_async_get_data_points_map_function(self):
+
+        def sum_results(inputIndex, inputObject, response, summation):
+            for point in response:
+                summation += point['value']
+            return summation
+
+        summation = self.client.batch_async_get_data_points(
+            [{'metric_id': 1, 'item_id': 2, 'region_id': 3, 'frequency_id': 4, 'source_id': 5}],
+            output_list=0,
+            map_result=sum_results
+        )
+
+        self.assertEqual(summation, 97680)
+
+    def test_batch_async_get_data_points_bad_request_error(self):
+        responses = self.client.batch_async_get_data_points([ERROR_SELECTION])
+        self.assertTrue(isinstance(responses[0], BatchError))
+
+    def test_batch_async_get_data_points_map_errors(self):
+
+        def raise_exception(idx, query, response, accumulator):
+            if isinstance(response, Exception):
+                raise response
+            accumulator[idx] = response
+            return accumulator
+
+        with self.assertRaises(Exception):
+            self.client.batch_async_get_data_points([ERROR_SELECTION],
+                                                    map_result=raise_exception)
+
+    def test_get_df(self):
+        self.client.add_single_data_series({
+            'metric_id': 1,
+            'item_id': 2,
+            'region_id': 3,
+            'frequency_id': 4,
+            'source_id': 5
+        })
+        df = self.client.get_df()
+        self.assertEqual(df.iloc[0]['start_date'].date(), date(2017, 1, 1))
+        self.assertEqual(df.iloc[0]['end_date'].date(), date(2017, 12, 31))
+        self.assertEqual(df.iloc[0]['value'], 40891)
+
+    def test_batch_async_rank_series_by_source(self):
+        list_of_ranked_series_lists = self.client.batch_async_rank_series_by_source(
+            [mock_data_series, mock_data_series]
+        )
+        # There were two inputs, so there should be two outputs:
+        self.assertEqual(len(list_of_ranked_series_lists), 2)
+        for series_list in list_of_ranked_series_lists:
+            # Not necessarily true, but true given the mock_rank_series_by_source() function:
+            self.assertEqual(len(series_list), len(mock_data_series))
+            for series in series_list:
+                self.assertTrue('metric_id' in series)
```

Comment:
Why assert that metric is in series and not other entity types?

Reply:
Could assert all of them, but I just picked metric_id and source_id randomly as a test set